language-design.md

```diff
@@ -38,18 +160,89 @@ How do the themes of _Growing a Language_ relate to the "sound lab" we did this
 
 **Response**
 
-
+Our main takeaway from _Growing a Language_ was that a good language should
+provide users with a small and useful set of building blocks.  From there, the
+programmer should be able to compose the available tools to suit their needs. We
+tried to use this mentality when adapting the Sound library to a more usable
+API.  First, we created a "Sound" class which contained all of the methods for
+editing sounds.  We tried to be consistent--each method took in a Sound (self)
+along with other parameters and returned a new Sound.  This way, the user could
+easily compose methods using calls like `mySound.reverse().overlay(otherSound)`.
+
+In addition, _Growing a Language_ showed the importance of keeping an API as 
+contained as possible, and only including options that would be relevant to 
+many users.  On pages 10 and 11, Steele shows many examples of data types, each
+increasingly obscure.  After describing them, Steele asks "So should we make
+'x' a type in the Java programming language?"  In the end, he concludes that
+language designers should not add these rare features because "It would not
+be fair to weigh down all programmers with the need to have or to learn all the
+words for all niche uses" (Steele, 12).  This idea ties with the theme of making
+a language/API exactly as small as it should be, and no smaller [Bloch, 2006]. 
+We tried to follow this maxim when designing our API for the sound lab. 
+All of the methods performed a single task, which allowed users to combine
+functions effectively.  We tried to keep the methods as segmented as possible,
+including options like "play", "overlay," "reverse," and "change speed".  We did
+not, for example, add a method that would both overlay two sounds and
+reverse them, as this would not have been useful for most users.
+
+Another idea that _Growing a Language_ emphasized was that users should be able
+to make their own functions that look like primitives of the language. 
+According to Steele, 
+> APL was designed by one man, a smart man—and I love APL—but it had a flaw that
+> I think has all but killed it: there was no way for a user to grow the
+> language in a smooth way (Steele, 6)
+
+In retrospect, we failed to consider this point when designing our Sound 
+API.  Our design involved creating a Sound class with two data members, the 
+sound samples and the sampling rate.  Our goal was to keep these data members 
+private and only expose users to the class's methods.  (Bloch includes this 
```

Comment:
The `_` prefix also prevents discovery by various tools for introspection (unless you try a bit harder). Additionally, getters and setters are unnecessary in Python because they use properties instead, i.e.

```python
class Foo:
    _bar = None

    @property
    def bar(self):
        if self._bar is None:
            self._bar = self.expensive_initialization_call()
        return self._bar

    @bar.setter
    def bar(self, value):
        if self.validate(value):
            self._bar = value
        else:
            raise ValueError("{} failed validation.".format(value))  

a = Foo()
print a.bar
a.bar = 42
```

It isn't that you can't or shouldn't use encapsulation in Python - it's just that there isn't a huge benefit to preventing users from accessing private values. If they choose to do so that's fine, but it's at their own risk.

Additionally, properties make Python much easier to extend than Java. With Java, the main reason we have `getX` and `setX` methods is because if we started at a `public int` but eventually wanted to calculate it instead we have a breaking interface change, thus we use `private int` and getters and setters so we don't have those breaking changes. With Python there are no breaking interface changes even if we want to make a property lazily evaluated, or provide extra validation logic, or whatever.